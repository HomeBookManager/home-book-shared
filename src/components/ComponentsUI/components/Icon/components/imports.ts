export { FC } from 'react';
export { omit } from 'lodash';

// styles
import '../styles/index.scss';

// types
export { TIconBasicProps } from '../types';
