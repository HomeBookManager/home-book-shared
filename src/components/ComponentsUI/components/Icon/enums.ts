export enum IconShape {
  iconDownOutlined = 'iconDownOutlined',
  iconLeftOutlined = 'iconLeftOutlined',
  iconRightOutlined = 'iconRightOutlined',
  iconUpOutlined = 'iconUpOutlined',
}

export enum IconModificatorsKeys {
  clickable = 'clickable',
  disabled = 'disabled',
  disabledHover = 'disabledHover',
  forcedHover = 'forcedHover',
  selected = 'selected',
}
