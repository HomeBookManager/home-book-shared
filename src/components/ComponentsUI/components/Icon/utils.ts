// @home-book
import { TObject } from 'home-book-types';

export const applyModificators = (className: string): TObject<string> => ({
  clickable: `${className}--clickable`,
  disabled: `${className}--disabled`,
  disabledHover: `${className}--disabled-hover`,
  forcedHover: `${className}--forced-hover`,
  selected: `${className}--selected`,
});
