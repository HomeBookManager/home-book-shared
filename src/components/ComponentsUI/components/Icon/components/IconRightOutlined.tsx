// others
import { FC, omit, TIconBasicProps } from './imports';

export const IconRightOutlined: FC<TIconBasicProps> = ({
  className = '',
  classNamePath = '',
  fill = '',
  ...restProps
}) => (
  <svg
    className={className}
    fill="none"
    height="16px"
    viewBox="64 64 896 896"
    width="16px"
    {...omit(restProps, 'stroke')}
  >
    <path
      className={classNamePath}
      d="M765.7 486.8L314.9 134.7A7.97 7.97 0 00302 141v77.3c0 4.9 2.3 9.6 6.1 12.6l360 281.1-360 281.1c-3.9 3-6.1 7.7-6.1 12.6V883c0 6.7 7.7 10.4 12.9 6.3l450.8-352.1a31.96 31.96 0 000-50.4z"
      fill={fill}
    ></path>
  </svg>
);

export default IconRightOutlined;
