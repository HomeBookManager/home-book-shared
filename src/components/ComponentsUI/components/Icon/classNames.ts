// types
import { IconShape } from './enums';
import { TIconShapes } from './types';

const classNamePrefix = 'Icon';
const classNameIconDownOutlined = `${classNamePrefix}DownOutlined`;
const classNameIconLeftOutlined = `${classNamePrefix}LeftOutlined`;
const classNameIconRightOutlined = `${classNamePrefix}RightOutlined`;
const classNameIconUpOutlined = `${classNamePrefix}UpOutlined`;

export const classNames: TIconShapes = {
  [IconShape.downOutlined]: classNameIconDownOutlined,
  [IconShape.leftOutlined]: classNameIconLeftOutlined,
  [IconShape.rightOutlined]: classNameIconRightOutlined,
  [IconShape.upOutlined]: classNameIconUpOutlined,
};
