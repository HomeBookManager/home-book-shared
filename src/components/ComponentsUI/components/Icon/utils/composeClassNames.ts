import { kebabCase } from 'lodash';

// types
import { IconModificatorsKeys } from '../enums';
import { TIconComposedClassNames, TIconModificators } from '../types';

// utils
import { enumToArray } from '../../../../../utils/transform/enumToArray';

export const applyModificators = (className: string): TIconModificators =>
  enumToArray<string>(IconModificatorsKeys).reduce(
    (obj, key) => ({
      ...obj,
      [key]: `${className}--${kebabCase(key)}`,
    }),
    {} as TIconModificators,
  );

export const composeClassNames = (
  componentName: string,
): TIconComposedClassNames => ({
  [componentName]: {
    modificators: applyModificators(componentName),
    name: componentName,
  },
});
