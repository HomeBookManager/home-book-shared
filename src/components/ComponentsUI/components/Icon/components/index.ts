import IconDownOutlined from './IconDownOutlined';
import IconLeftOutlined from './IconLeftOutlined';
import IconRightOutlined from './IconRightOutlined';
import IconUpOutlined from './IconUpOutlined';

export const Icons = {
  IconDownOutlined,
  IconLeftOutlined,
  IconRightOutlined,
  IconUpOutlined,
};

export default Icons;
