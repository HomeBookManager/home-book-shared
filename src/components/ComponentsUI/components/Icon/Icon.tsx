import cx from 'classnames';
import { FC, Ref, forwardRef } from 'react';
import { kebabCase } from 'lodash';

// @home-book
import { Theme } from 'home-book-types';

// components
import E2EDataAttribute from '../../../E2EDataAttributes/E2EDataAttribute';
import withRefreshReference from '../../../../shared/WithRefreshReference/withRefreshReference';

// hooks
import { useTheme } from '../../../../hooks';

// styles
import './styles/index.scss';

// types
import { E2EAttribute } from '../../../E2EDataAttributes/enums';
import { TIconBasicProps } from './types';

// utils
import { composeClassNames } from './utils/composeClassNames';

export type TIconProps = TIconBasicProps & {
  fillDark?: string;
  iconComponent: FC<TIconBasicProps>;
  ignoreDefaultStyles?: boolean;
  ref?: Ref<SVGSVGElement>;
  selected?: boolean;
  strokeDark?: string;
};

export const Icon = forwardRef<SVGSVGElement, TIconProps>(
  (
    {
      className = '',
      clickable = false,
      disabled = false,
      disabledFocus = false,
      disabledHover = false,
      e2eAttribute = E2EAttribute.icon,
      e2eValue = '',
      fill = '',
      fillDark = '',
      forcedHover = false,
      iconComponent: IconComponent,
      ignoreDefaultStyles = false,
      selected = false,
      stroke = '',
      strokeDark = '',
      tabIndex = -1,
      ...restProps
    },
    ref,
  ) => {
    const componentName = IconComponent.name;

    const withDefaultStyles =
      !ignoreDefaultStyles && !fill && !fillDark && !stroke && !strokeDark;

    const { classNamesWithTheme, theme } = useTheme(
      composeClassNames(componentName),
    );

    return (
      <E2EDataAttribute
        type={e2eAttribute}
        value={e2eValue || kebabCase(componentName)}
      >
        <IconComponent
          className={
            withDefaultStyles
              ? cx(
                  className,
                  classNamesWithTheme[componentName].name,
                  {
                    [classNamesWithTheme[componentName].modificators.clickable]:
                      clickable,
                  },
                  {
                    [classNamesWithTheme[componentName].modificators.disabled]:
                      disabled,
                  },
                  {
                    [classNamesWithTheme[componentName].modificators
                      .disabledHover]: disabledHover,
                  },
                  {
                    [classNamesWithTheme[componentName].modificators
                      .forcedHover]: forcedHover,
                  },
                  {
                    [classNamesWithTheme[componentName].modificators.selected]:
                      selected,
                  },
                )
              : className
          }
          fill={theme === Theme.dark ? fillDark : fill}
          ref={ref}
          stroke={theme === Theme.dark ? strokeDark : stroke}
          {...(disabledFocus ? {} : { tabIndex })}
          {...restProps}
        />
      </E2EDataAttribute>
    );
  },
);

export default withRefreshReference(Icon);
