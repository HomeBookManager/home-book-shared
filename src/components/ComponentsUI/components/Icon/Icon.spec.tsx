import { render } from '@testing-library/react';

// @home-book
import { Theme } from 'home-book-types';

// components
import Icon from './Icon';
import { IconUpOutlined } from './components';

// core
import { ContextProvider } from '../../../../core/ContextProvider/ContextProvider';

// others
import { classNames as classNamesIcon } from './classNames';

// types
import { E2EAttribute } from '../../../E2EDataAttributes/enums';
import { IconShape } from './enums';

// utils
import { composeClassNames } from './utils/composeClassNames';
import { getByE2EAttribute } from '../../../../tests/testHelpers';
import { getDataTestAttribute } from '../../../E2EDataAttributes/utils';

const iconComponent = IconUpOutlined;
const iconShape = IconShape.upOutlined;

const className = 'className';
const classNamePath = 'classNamePath';

const classNameIconShape = classNamesIcon[iconShape];
const classNames = composeClassNames(iconShape);

describe('Icon behaviors', () => {
  it('should not apply default styles when all specific props passed', () => {
    // mock
    const fill = '#000000';
    const fillDark = '#000000';
    const stroke = '#000000';
    const strokeDark = '#000000';

    //before
    const { container } = render(
      <ContextProvider>
        <Icon
          fill={fill}
          fillDark={fillDark}
          iconComponent={iconComponent}
          iconShape={iconShape}
          ignoreDefaultStyles
          stroke={stroke}
          strokeDark={strokeDark}
        />
      </ContextProvider>,
    );

    // result
    expect(getByE2EAttribute(container, E2EAttribute.icon)).not.toHaveClass(
      classNames[classNameIconShape].name,
    );
  });

  it('should not apply default styles when ignoreDefaultStyles passed', () => {
    //before
    const { container } = render(
      <ContextProvider>
        <Icon
          iconComponent={iconComponent}
          iconShape={iconShape}
          ignoreDefaultStyles
        />
      </ContextProvider>,
    );

    // result
    expect(getByE2EAttribute(container, E2EAttribute.icon)).not.toHaveClass(
      classNames[classNameIconShape].name,
    );
  });

  it('should not apply default styles when fill passed', () => {
    // mock
    const fill = '#000000';

    //before
    const { container } = render(
      <ContextProvider>
        <Icon fill={fill} iconComponent={iconComponent} iconShape={iconShape} />
      </ContextProvider>,
    );

    // result
    expect(getByE2EAttribute(container, E2EAttribute.icon)).not.toHaveClass(
      classNames[classNameIconShape].name,
    );
  });

  it('should not apply default styles when fillDark passed', () => {
    // mock
    const fillDark = '#000000';

    //before
    const { container } = render(
      <ContextProvider>
        <Icon
          fillDark={fillDark}
          iconComponent={iconComponent}
          iconShape={iconShape}
        />
      </ContextProvider>,
    );

    // result
    expect(getByE2EAttribute(container, E2EAttribute.icon)).not.toHaveClass(
      classNames[classNameIconShape].name,
    );
  });

  it('should not apply default styles when stroke passed', () => {
    // mock
    const stroke = '#000000';

    //before
    const { container } = render(
      <ContextProvider>
        <Icon
          iconComponent={iconComponent}
          iconShape={iconShape}
          stroke={stroke}
        />
      </ContextProvider>,
    );

    // result
    expect(getByE2EAttribute(container, E2EAttribute.icon)).not.toHaveClass(
      classNames[classNameIconShape].name,
    );
  });

  it('should not apply default styles when strokeDark passed', () => {
    // mock
    const strokeDark = '#000000';

    //before
    const { container } = render(
      <ContextProvider>
        <Icon
          iconComponent={iconComponent}
          iconShape={iconShape}
          strokeDark={strokeDark}
        />
      </ContextProvider>,
    );

    // result
    expect(getByE2EAttribute(container, E2EAttribute.icon)).not.toHaveClass(
      classNames[classNameIconShape].name,
    );
  });
});

describe('Icon props', () => {
  it('should pass className', () => {
    //before
    const { container } = render(
      <ContextProvider>
        <Icon
          className={className}
          iconComponent={iconComponent}
          iconShape={iconShape}
        />
      </ContextProvider>,
    );

    // result
    expect(container.querySelector(`.${className}`)).toHaveClass(className);
  });

  it('should pass classNamePath', () => {
    //before
    const { container } = render(
      <ContextProvider>
        <Icon
          classNamePath={classNamePath}
          iconComponent={iconComponent}
          iconShape={iconShape}
        />
      </ContextProvider>,
    );

    // result
    expect(container.querySelector(`.${classNamePath}`)).toHaveClass(
      classNamePath,
    );
  });

  it('should pass clickable', () => {
    //before
    const { container } = render(
      <ContextProvider>
        <Icon clickable iconComponent={iconComponent} iconShape={iconShape} />
      </ContextProvider>,
    );

    // result
    expect(getByE2EAttribute(container, E2EAttribute.icon)).toHaveClass(
      classNames[classNameIconShape].modificators.clickable,
    );
  });

  it('should pass disabled', () => {
    //before
    const { container } = render(
      <ContextProvider>
        <Icon disabled iconComponent={iconComponent} iconShape={iconShape} />
      </ContextProvider>,
    );

    // result
    expect(getByE2EAttribute(container, E2EAttribute.icon)).toHaveClass(
      classNames[classNameIconShape].modificators.disabled,
    );
  });

  it('should pass disabledFocus', () => {
    //before
    const { container } = render(
      <ContextProvider>
        <Icon
          disabledFocus
          iconComponent={iconComponent}
          iconShape={iconShape}
        />
      </ContextProvider>,
    );

    // result
    expect(getByE2EAttribute(container, E2EAttribute.icon)).not.toHaveAttribute(
      'tabindex',
    );
  });

  it('should pass disabledHover', () => {
    //before
    const { container } = render(
      <ContextProvider>
        <Icon
          disabledHover
          iconComponent={iconComponent}
          iconShape={iconShape}
        />
      </ContextProvider>,
    );

    // result
    expect(getByE2EAttribute(container, E2EAttribute.icon)).toHaveClass(
      classNames[classNameIconShape].modificators.disabledHover,
    );
  });

  it('should pass e2eAttribute', () => {
    // before
    const { container } = render(
      <ContextProvider>
        <Icon
          e2eAttribute={E2EAttribute.icon}
          iconComponent={iconComponent}
          iconShape={iconShape}
        />
      </ContextProvider>,
    );

    // result
    expect(getByE2EAttribute(container, E2EAttribute.icon)).toHaveAttribute(
      getDataTestAttribute(E2EAttribute.icon),
    );
  });

  it('should pass e2eValue', () => {
    // mock
    const e2eValue = 'e2eValue';

    // before
    const { container } = render(
      <ContextProvider>
        <Icon
          e2eValue={e2eValue}
          iconComponent={iconComponent}
          iconShape={iconShape}
        />
      </ContextProvider>,
    );

    // result
    expect(getByE2EAttribute(container, E2EAttribute.icon)).toHaveAttribute(
      getDataTestAttribute(E2EAttribute.icon),
      e2eValue,
    );
  });

  it('should pass fill', () => {
    // mock
    const fill = '#000000';

    //before
    const { container } = render(
      <ContextProvider>
        <Icon
          classNamePath={classNamePath}
          fill={fill}
          iconComponent={iconComponent}
          iconShape={iconShape}
        />
      </ContextProvider>,
    );

    // result
    expect(container.querySelector(`.${classNamePath}`)).toHaveAttribute(
      'fill',
      fill,
    );
  });

  it('should pass fillDark', () => {
    // mock
    const fillDark = '#000000';

    //before
    const { container } = render(
      <ContextProvider theme={Theme.dark}>
        <Icon
          classNamePath={classNamePath}
          fillDark={fillDark}
          iconComponent={iconComponent}
          iconShape={iconShape}
        />
      </ContextProvider>,
    );

    // result
    expect(container.querySelector(`.${classNamePath}`)).toHaveAttribute(
      'fill',
      fillDark,
    );
  });

  it('should pass forcedHover', () => {
    //before
    const { container } = render(
      <ContextProvider>
        <Icon forcedHover iconComponent={iconComponent} iconShape={iconShape} />
      </ContextProvider>,
    );

    // result
    expect(getByE2EAttribute(container, E2EAttribute.icon)).toHaveClass(
      classNames[classNameIconShape].modificators.forcedHover,
    );
  });

  it('should pass iconComponent', () => {
    //before
    const { container } = render(
      <ContextProvider>
        <Icon iconComponent={iconComponent} iconShape={iconShape} />
      </ContextProvider>,
    );

    // result
    expect(getByE2EAttribute(container, E2EAttribute.icon).tagName).toBe('svg');
  });

  it('should pass iconShape', () => {
    //before
    const { container } = render(
      <ContextProvider>
        <Icon iconComponent={iconComponent} iconShape={iconShape} />
      </ContextProvider>,
    );

    // result
    expect(getByE2EAttribute(container, E2EAttribute.icon)).toHaveClass(
      classNames[classNameIconShape].name,
    );
  });

  it('should pass ignoreDefaultStyles', () => {
    //before
    const { container } = render(
      <ContextProvider>
        <Icon
          iconComponent={iconComponent}
          iconShape={iconShape}
          ignoreDefaultStyles
        />
      </ContextProvider>,
    );

    // result
    expect(getByE2EAttribute(container, E2EAttribute.icon)).not.toHaveClass(
      classNames[classNameIconShape].name,
    );
  });

  it('should pass selected', () => {
    //before
    const { container } = render(
      <ContextProvider>
        <Icon iconComponent={iconComponent} iconShape={iconShape} selected />
      </ContextProvider>,
    );

    // result
    expect(getByE2EAttribute(container, E2EAttribute.icon)).toHaveClass(
      classNames[classNameIconShape].modificators.selected,
    );
  });

  it('should pass stroke', () => {
    // mock
    const stroke = '#000000';

    //before
    const { container } = render(
      <ContextProvider>
        <Icon
          classNamePath={classNamePath}
          iconComponent={iconComponent}
          iconShape={iconShape}
          stroke={stroke}
        />
      </ContextProvider>,
    );

    // result
    expect(container.querySelector(`.${classNamePath}`)).not.toHaveAttribute(
      'stroke',
      stroke,
    );
  });

  it('should pass strokeDark', () => {
    // mock
    const strokeDark = '#000000';

    //before
    const { container } = render(
      <ContextProvider theme={Theme.dark}>
        <Icon
          classNamePath={classNamePath}
          iconComponent={iconComponent}
          iconShape={iconShape}
          strokeDark={strokeDark}
        />
      </ContextProvider>,
    );

    // result
    expect(container.querySelector(`.${classNamePath}`)).not.toHaveAttribute(
      'stroke',
      strokeDark,
    );
  });
});

describe('Icon snapshots', () => {
  it('should render Icon', () => {
    //before
    const { asFragment } = render(
      <ContextProvider>
        <Icon iconComponent={iconComponent} iconShape={iconShape} />
      </ContextProvider>,
    );

    // result
    expect(asFragment()).toMatchSnapshot();
  });
});
