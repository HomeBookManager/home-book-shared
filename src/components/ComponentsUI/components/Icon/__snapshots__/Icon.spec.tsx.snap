// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Icon snapshots should render Icon 1`] = `
<DocumentFragment>
  <svg
    class="IconUpOutlined"
    data-test-icon="icon-up-outlined"
    fill="none"
    height="16px"
    tabindex="-1"
    viewBox="64 64 896 896"
    width="16px"
  >
    <path
      class=""
      d="M890.5 755.3L537.9 269.2c-12.8-17.6-39-17.6-51.7 0L133.5 755.3A8 8 0 00140 768h75c5.1 0 9.9-2.5 12.9-6.6L512 369.8l284.1 391.6c3 4.1 7.8 6.6 12.9 6.6h75c6.5 0 10.3-7.4 6.5-12.7z"
      fill=""
    />
  </svg>
</DocumentFragment>
`;
